### PR TITLE
Remove legacy kolena.{classification,detection,fr} workflows

### PR DESCRIPTION
**Pull request by sandalns**
### Linked issue(s):
Fixes KOL-4536

### What change does this PR introduce and why?
Removes classification, detection and fr
### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -30,34 +30,17 @@ repos:
     -   id: insert-license
         files: ^(kolena|tests|examples)/.*\.py$
         args: [--license-filepath, "LICENSE_HEADER"]
--   repo: https://github.com/psf/black
```
Comment (sandalns) [inline .pre-commit-config.yaml]: This is temporary to speed up iteration

File: kolena/_utils/cli.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 import json
 import os
 import sys
+from typing import Any
```
Comment (sandalns) [inline kolena/_utils/cli.py:18]: There is likely a better way for this, flagging to check later

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -30,34 +30,17 @@ repos:
...
-        files: ^examples/dataset/
-        additional_dependencies: [ types-all, "pydantic<2.0" ]
-      - id: mypy
-        name: mypy-examples-workflow
-        files: ^examples/workflow/
+        exclude: |
+            (?x)^(
+                examples/.*|
+                tests/.* |
+                kolena/_experimental/.*
```
Comment (sandalns) [inline .pre-commit-config.yaml]: After we resolve the remaing issues, we can try removing the exclude on `kolena/_experimental/.*` (adds another 100 mypy issues). The tests exclude will be harder to remove (may require a mypy upgrade).

File: kolena/_utils/krequests.py [hunk 4]
```diff
@@ -143,21 +144,22 @@ def delete(endpoint_path: str, api_version: int = DEFAULT_API_VERSION, **kwargs:
 def raise_for_status(response: requests.Response) -> None:
```
Comment (sandalns) [inline kolena/_utils/krequests.py]: Wondering if it is safer to just add a type ignore here instead of the changes to the errors for now. Want to limit the chances of breakages and these are widely used

File: .circleci/continue_config.yml [hunk 5]
```diff
@@ -424,7 +415,7 @@ workflows:
       - example-test-workflow:
           matrix:
             parameters:
-              subproject: [ keypoint_detection, age_estimation, object_detection_3d, classification, question_answering, face_recognition_11, speaker_diarization, automatic_speech_recognition ]
```
Comment (ly-kolena) [inline .circleci/continue_config.yml]: Why is this removed?

File: kolena/_api/v1/core.py [hunk 6]
```diff
@@ -29,7 +30,7 @@ class CreateRequest:
         name: str
         metadata: Dict[str, Any]
         workflow: str
-        tags: Optional[List[str]] = None
+        tags: Optional[Union[List[str], Set[str]]] = None
```
Comment (ly-kolena) [inline kolena/_api/v1/core.py]: Why is this change needed?

File: kolena/_utils/datatypes.py [hunk 7]
```diff
@@ -80,11 +80,11 @@ def _double_under(input: str) -> bool:
 def _allow_extra(cls: Type[T]) -> bool:
     # `pydantic.dataclasses.is_built_in_dataclass` would have false-positive when a stdlib-dataclass decorated
     # class extends a pydantic dataclass
-    return "__pydantic_model__" in vars(cls) and cls.__pydantic_model__.Config.extra == Extra.allow
+    return "__pydantic_model__" in vars(cls) and cls.__pydantic_model__.Config.extra == Extra.allow  # type: ignore
 
 
 # used to track data_type string -> TypedDataObject
-_DATA_TYPE_MAP = {}
+_DATA_TYPE_MAP: Dict[Any, Any] = {}
```
Comment (ly-kolena) [inline kolena/_utils/datatypes.py]: Can this be scoped to `dict[str, TypedDataObject]` ?

File: kolena/_utils/datatypes.py [hunk 8]
```diff
@@ -97,7 +97,7 @@ def _get_data_type(name: str) -> Optional[Type["TypedDataObject"]]:
 
 
 # used for TypedBaseDataObject to register themselves to be used in dataclass extra fields deserialization
-def _register_data_type(cls) -> None:
+def _register_data_type(cls: Any) -> None:
```
Comment (ly-kolena) [inline kolena/_utils/datatypes.py]: Minor similar question, can this be scoped to TypedDataObject class type?

File: kolena/_utils/krequests.py [hunk 9]
```diff
@@ -143,21 +144,22 @@ def delete(endpoint_path: str, api_version: int = DEFAULT_API_VERSION, **kwargs:
...
+        raise NameConflictError(response=response)
     if response.status_code == HTTPStatus.UNPROCESSABLE_ENTITY:
         raise IncorrectUsageError(response.content)
 
     try:
         response.raise_for_status()
     except HTTPError:
-        raise RemoteError(f"{response.text} ({response.elapsed.total_seconds():0.5f} seconds elapsed)")
+        error_message = f"{response.text} ({response.elapsed.total_seconds():0.5f} seconds elapsed)"
+        raise RemoteError(response=error_message)  # type: ignore
```
Comment (ly-kolena) [inline kolena/_utils/krequests.py]: Is this right? If the change is for `HttpError`, it is expecting response to be a response object. Maybe 
`RemoteError(error_message, response=response)`?

File: kolena/_utils/krequests.py [hunk 10]
```diff
@@ -143,21 +144,22 @@ def delete(endpoint_path: str, api_version: int = DEFAULT_API_VERSION, **kwargs:
 def raise_for_status(response: requests.Response) -> None:
     if response.status_code == HTTPStatus.UNAUTHORIZED:
         # HTTP 401 is "unauthorized" but used as "unauthenticated"
-        raise UnauthenticatedError(response.content)
+        raise UnauthenticatedError(response=response)
```
Comment (ly-kolena) [inline kolena/_utils/krequests.py]: Can you double check the impact of error reporting on this change, for example console output or stack trace? What would it look like before and after?

File: kolena/dataset/dataset.py [hunk 11]
```diff
@@ -71,7 +71,7 @@ class DatapointType(str, Enum):
     VIDEO = "DATAPOINT/VIDEO"
 
     @classmethod
-    def has_value(cls, item) -> bool:
+    def has_value(cls, item) -> bool:  # type: ignore
```
Comment (ly-kolena) [inline kolena/dataset/dataset.py]: Would it do to use `item: Union[str, "DatapointType"]`?

File: kolena/dataset/dataset.py [hunk 12]
```diff
@@ -91,7 +91,7 @@ def _dataobject_type(obj: TypedDataObject) -> str:
 
 def _get_datapoint_type(mimetype_str: str) -> str:
     main_type, sub_type = mimetype_str.split("/")
-    return _DATAPOINT_TYPE_MAP.get(mimetype_str, None) or _DATAPOINT_TYPE_MAP.get(main_type, None)
+    return _DATAPOINT_TYPE_MAP.get(mimetype_str, None) or _DATAPOINT_TYPE_MAP.get(main_type, None)  # type: ignore
```
Comment (ly-kolena) [inline kolena/dataset/dataset.py]: Should we update the return type to `Optional[str]` instead?

File: tests/integration/workflow/test_model.py [hunk 13]
```diff
@@ -82,13 +80,6 @@ def test__load_all() -> None:
     assert model.load_all(tags={"does_not_exist"}) == []
 
 
-def test__load__mismatching_workflows() -> None:
```
Comment (ly-kolena) [inline tests/integration/workflow/test_model.py:85]: Alternative is register another generic workflow to keep validations of this kind.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (ly-kolena) [reply .pre-commit-config.yaml]: Can you elaborate on this change? Why are `mypy-examples-XXX` removed?

File: kolena/_utils/instrumentation.py [hunk 14]
```diff
@@ -128,11 +128,14 @@ def wrapper(*args: Any, **kwargs: Any) -> Any:
                 event_metadata["response_error"] = e.__class__.__name__
                 raise e
             finally:
-                event_metadata["duration"] = round((datetime.datetime.now() - start_time).total_seconds(), 3)
+                event_metadata["duration"] = round(
```
Comment (ly-kolena) [inline kolena/_utils/instrumentation.py]: The 2 ignores can be resolved by specifying typing for `event_metadata` above:

```
event_metadata: Dict[str, Union[str, int, float, bool, None]] = {}
```

File: kolena/_utils/krequests.py [hunk 15]
```diff
@@ -83,7 +84,7 @@ def _with_default_kwargs(**kwargs: Any) -> Dict[str, Any]:
     )
     # allow requests to override Content-Type as needed by for example file uploads
     if "Content-Type" in kwargs.get("headers", {}):
-        default_headers["Content-Type"] = kwargs.get("headers")["Content-Type"]
+        default_headers["Content-Type"] = kwargs.get("headers")["Content-Type"]  # type: ignore
```
Comment (ly-kolena) [inline kolena/_utils/krequests.py]: this ignore can be resolved by

```
    kw_headers = kwargs.get("headers", {})
    if "Content-Type" in kw_headers:
        default_headers["Content-Type"] = kw_headers["Content-Type"]
```

File: kolena/_utils/batched_load.py [hunk 16]
```diff
@@ -112,10 +112,10 @@ def concat(dfs: Iterable[pd.DataFrame], df_class: Type[DFType]) -> DFType:
         return df_class(df)
```
Comment (ly-kolena) [inline kolena/_utils/batched_load.py:110]: There's a mypy error on this line, I think it can be resolved by `return df`?

File: kolena/_utils/batched_load.py [hunk 17]
```diff
@@ -128,7 +128,7 @@ def iter_data(
         init_request: API.BaseInitDownloadRequest,
         endpoint_path: str,
         df_class: Optional[Type[DFType]],
-        endpoint_api_version: int = DEFAULT_API_VERSION,
+        endpoint_api_version: str = DEFAULT_API_VERSION,
     ) -> Iterator[DFType]:
         kreq = krequests if endpoint_api_version == API_V1 else krequests_v2
         with kreq.put(
```
Comment (ly-kolena) [inline kolena/_utils/batched_load.py]: mypy has issue with module as variable, which they don't seem to intend to support. Unfortunately have to `#type: ignore` on this kind of usage in this file.

File: kolena/workflow/test_run.py [hunk 18]
```diff
@@ -126,11 +126,15 @@ def __init__(
...
+        self.configurations = self.evaluator.configurations if is_evaluator_class else configurations  # type: ignore
         self.reset = reset
 
         evaluator_display_name = (
-            None if evaluator is None else evaluator.display_name() if is_evaluator_class else evaluator.__name__
+            None
+            if evaluator is None
+            else evaluator.display_name()
+            if is_evaluator_class
+            else evaluator.__name__  # type: ignore
```
Comment (ly-kolena) [inline kolena/workflow/test_run.py]: I think this actually caught a potentially valid issue, not all callable has `__name__`. Proposed resolution

```
def get_evaluator_display_name(evaluator: Optional[Union[Evaluator, BasicEvaluatorFunction]]) -> Optional[str]:
    if not evaluator:
        return None
    if isinstance(evaluator, Evaluator):
        return evaluator.display_name()
    else:
        return getattr(evaluator, "__name__", repr(evaluator))
```
and use
```
evaluator_display_name = get_evaluator_display_name(evaluator)
```

File: .circleci/continue_config.yml (hunk 5 above)
Comment (munkyshi) [reply .circleci/continue_config.yml]: Agreed, this looks like it should be for the pre-built workflow impl of classification, not for the legacy classification workflow.

File: kolena/workflow/test_run.py [hunk 19]
```diff
@@ -370,19 +374,24 @@ def process_results(results: Optional[EvaluationResults], config: Optional[Evalu
...
+                inferences,
+                test_case_test_samples,
+            )  # type: ignore
             process_results(evaluation_results, None)
 
         log.info("uploading test case metrics")
         self._upload_test_case_metrics(test_case_metrics)
         log.info("uploading test case plots")
-        self._upload_test_case_plots(test_case_plots)
+        self._upload_test_case_plots(test_case_plots)  # type: ignore
```
Comment (ly-kolena) [inline kolena/workflow/test_run.py]: Nit: for these 2 ignores, seems like alternative could be

```
    def _upload_test_case_plots(
        self,
        plots: Mapping[int, Mapping[Optional[EvaluatorConfiguration], Optional[List[Plot]]]],
    ) -> None:
      ...

    def _upload_test_suite_metrics(
        self,
        metrics: Mapping[Optional[EvaluatorConfiguration], Optional[MetricsTestSuite]],
    ) -> None:
```

File: kolena/_utils/state.py [hunk 20]
```diff
@@ -159,6 +159,7 @@ def get_token(
     proxies: Optional[Dict[str, str]] = None,
 ) -> API.ValidateResponse:
     base_url = base_url or get_client_state().base_url
+    assert base_url is not None
```
Comment (munkyshi) [inline kolena/_utils/state.py]: We should explicitly throw an error here -- assertion errors are generally more likely to be caught / skipped over.

File: kolena/initialize.py [hunk 21]
```diff
@@ -120,7 +120,7 @@ def initialize(
             category=DeprecationWarning,
             stacklevel=2,
         )
-
+    assert api_token is not None
```
Comment (munkyshi) [inline kolena/initialize.py:123]: Same comment re asserting

File: kolena/initialize.py [hunk 22]
```diff
@@ -145,11 +145,12 @@ def initialize(
         log.info(f"connected to {get_platform_url()}")
 
 
-def _find_token() -> str:
+def _find_token() -> Optional[str]:
     if KOLENA_TOKEN_ENV in os.environ:
         return os.environ[KOLENA_TOKEN_ENV]
 
     hostname = urlparse(state._get_api_base_url()).hostname
+    assert hostname is not None
```
Comment (munkyshi) [inline kolena/initialize.py]: Same comment re asserting

File: kolena/workflow/test_case.py [hunk 23]
```diff
@@ -444,7 +446,7 @@ def init_many(
         bulk_response = from_dict(data_class=CoreAPI.BulkProcessResponse, data=response.json())
 
         test_cases = []
-        statuses = defaultdict(int)
+        statuses: DefaultDict[Any, int] = defaultdict(int)
```
Comment (munkyshi) [inline kolena/workflow/test_case.py]: Can this be `DefaultDict[BulkProcessStatus, int]`?

File: kolena/workflow/visualization/_activation_map.py [hunk 24]
```diff
@@ -56,7 +56,7 @@ def blue(self, intensity: np.uint8) -> np.uint8:
         """
         raise NotImplementedError
 
-    def alpha(self, intensity: np.uint8) -> np.uint8:
+    def alpha(self, intensity: np.uint8) -> Union[np.uint8, int]:
```
Comment (munkyshi) [inline kolena/workflow/visualization/_activation_map.py]: Should this just be `int` (since it's a superset of `np.uint8`)?

File: tests/integration/workflow/test_model.py (hunk 13 above)
Comment (munkyshi) [reply tests/integration/workflow/test_model.py:85]: If we merge with this removed, we should track follow ups to re-add / re-enable tests in a linear issue.

File: tests/unit/test_initialize.py [hunk 25]
```diff
@@ -169,11 +168,6 @@ def test__initialize__token_missing() -> None:
             kolena.initialize()
 
 
-def test__uninitialized_usage() -> None:
-    with pytest.raises(UninitializedError):
-        kolena.fr.Model.create("test", {})
```
Comment (munkyshi) [inline tests/unit/test_initialize.py]: We should still test this -- it's functionality that's relevant still.

File: docs/reference/legacy/index.md
Comment (gordonhart) [inline docs/reference/legacy/index.md:1]: We should leave a pointer somewhere (not necessarily here) to [legacy-docs.kolena.io](https://legacy-docs.kolena.io/) and client version 0.x.

File: .circleci/continue_config.yml (hunk 5 above)
Comment (gordonhart) [reply .circleci/continue_config.yml]: Re-added :+1:

File: kolena/_api/v1/core.py (hunk 6 above)
Comment (gordonhart) [reply kolena/_api/v1/core.py]: Removed and addressed need for this change elsewhere.

File: kolena/_utils/datatypes.py (hunk 7 above)
Comment (gordonhart) [reply kolena/_utils/datatypes.py]: Done

File: kolena/_utils/datatypes.py (hunk 8 above)
Comment (gordonhart) [reply kolena/_utils/datatypes.py]: Done

File: kolena/dataset/dataset.py (hunk 11 above)
Comment (gordonhart) [reply kolena/dataset/dataset.py]: Not quite but `Any` works and is probably the best signature here. Updated

File: kolena/_utils/instrumentation.py (hunk 14 above)
Comment (gordonhart) [reply kolena/_utils/instrumentation.py]: Done

File: kolena/_utils/krequests.py (hunk 15 above)
Comment (gordonhart) [reply kolena/_utils/krequests.py]: 🙇

File: .circleci/config.yml [hunk 26]
```diff
@@ -42,7 +36,7 @@ jobs:
           name: Generate config
           command: |
             export INT_MATRIX=$(jq -c 'if (.all) then
-              (["detection", "fr", "classification", "workflow"])
+              (["workflow"])
```
Comment (gordonhart) [inline .circleci/config.yml:39]: We should be able to remove complexity in a follow-up now that there is only the `workflow` option here.

File: kolena/_utils/batched_load.py (hunk 16 above)
Comment (gordonhart) [reply kolena/_utils/batched_load.py:110]: Resolved by cleaning up `LoadableDataFrame` class hierarchy.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (gordonhart) [reply .pre-commit-config.yaml]: Added `mypy-examples-dataset` back. We'll add `mypy-examples-workflow` back in a follow-up — many typing issues there were unmasked by this PR

File: kolena/workflow/test_run.py (hunk 18 above)
Comment (gordonhart) [reply kolena/workflow/test_run.py]: Updated to remove type ignore

File: kolena/workflow/test_case.py (hunk 23 above)
Comment (gordonhart) [reply kolena/workflow/test_case.py]: Done

File: tests/integration/workflow/test_model.py (hunk 13 above)
Comment (gordonhart) [reply tests/integration/workflow/test_model.py:85]: Added back

File: tests/unit/test_initialize.py (hunk 25 above)
Comment (gordonhart) [reply tests/unit/test_initialize.py]: Added back

File: kolena/_utils/state.py (hunk 20 above)
Comment (gordonhart) [reply kolena/_utils/state.py]: These assertions exist to make mypy happy — due to other conditions, `base_url` here and other variables elsewhere where `assert` is added in this PR are non-null, meaning users should never see AssertionErrors. Agree that this is less than ideal, but it's less annoying to add these assertions for the time being than to refactor types in this PR

File: kolena/_utils/krequests.py (hunk 9 above)
Comment (gordonhart) [reply kolena/_utils/krequests.py]: Reverted

File: kolena/_utils/krequests.py (hunk 10 above)
Comment (gordonhart) [reply kolena/_utils/krequests.py]: Reverted

File: kolena/workflow/test_run.py (hunk 19 above)
Comment (gordonhart) [reply kolena/workflow/test_run.py]: Updated, thanks for the suggestion :+1:

File: kolena/workflow/visualization/_activation_map.py (hunk 24 above)
Comment (gordonhart) [reply kolena/workflow/visualization/_activation_map.py]: Updated. This file may need another look at some point (types are off, boundary between normal int/float and uint8 is not handled well) but passes mypy and unit tests for now.

**Review by ly-kolena — APPROVED** [review]

**Review by gordonhart — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 27]
```diff
@@ -45,20 +45,22 @@ repos:
...
         additional_dependencies: [ types-all, "pydantic<2.0" ]
       - id: mypy
         name: mypy-examples-dataset
         files: ^examples/dataset/
         additional_dependencies: [ types-all, "pydantic<2.0" ]
-      - id: mypy
-        name: mypy-examples-workflow
-        files: ^examples/workflow/
-        additional_dependencies: [ types-all, "pydantic<2.0" ]
+      # TODO: enable
```
Comment (munkyshi) [inline .pre-commit-config.yaml:59]: https://linear.app/kolena/issue/KOL-4575/restore-mypy-checks-for-examples
